src/main/java/com/aws/greengrass/mqtt/bridge/util/TopicMatcher.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.mqtt.bridge.util;
+
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Implementation adapted from <a href="https://github.com/eclipse/paho.mqtt.java/blob/f4e0db802a4433645ef011e711646a09ec9fae89/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/MqttTopic.java#L253-L321">paho</a>.
+ */
+@SuppressWarnings("PMD.UseVarargs")
+public final class TopicMatcher {
+    private static final String TOPIC_LEVEL_SEPARATOR = "/";
+    private static final String MULTI_LEVEL_WILDCARD = "#";
+    private static final String SINGLE_LEVEL_WILDCARD = "+";
+    private static final String MULTI_LEVEL_WILDCARD_PATTERN = TOPIC_LEVEL_SEPARATOR + MULTI_LEVEL_WILDCARD;
+    private static final String TOPIC_WILDCARDS = MULTI_LEVEL_WILDCARD + SINGLE_LEVEL_WILDCARD;
+    private static final char[] TOPIC_WILDCARDS_ARR = TopicMatcher.TOPIC_WILDCARDS.toCharArray();
+    private static final int MIN_TOPIC_LEN = 1;
+    private static final int MAX_TOPIC_LEN = 65_535;
+    private static final char NUL = '\u0000';
+    private static final int INDEX_NOT_FOUND = -1;
+
+    private TopicMatcher() {
+    }
+
+    /**
+     * True if topic matches the given filter.
+     *
+     * @param topicFilter topic filter
+     * @param topicName   topic name
+     * @return true if topic matches the filter
+     * @throws IllegalArgumentException if filter is invalid
+     */
+    public static boolean isMatched(String topicFilter, String topicName) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.